algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -458,6 +458,59 @@ case class CMSItem[K](item: K, override val totalCount: Long, override val param
 }
 
 /**
+ * A sparse Count-Min sketch structure, used for situations where the key is highly skewed.
+ */
+case class SparseCMS[K](exactCountTable: Map[K, Long],
+  override val totalCount: Long,
+  override val params: CMSParams[K]) extends CMS[K](params) {
+
+  override def +(x: K, count: Long): CMS[K] = {
+    val currentCount = exactCountTable.getOrElse(x, 0L)
+    val newTable = exactCountTable.updated(x, currentCount + count)
+    if (newTable.size < width * depth) {
+      // still sparse
+      SparseCMS(newTable, totalCount + count, params)
+    } else {
+      // Create new CMSInstace
+      var cms = CMSInstance[K](params)
+      newTable.foreach {
+        case (x, count) =>
+          cms = cms + (x, count)
+      }
+      cms
+    }
+  }
+
+  override def ++(other: CMS[K]): CMS[K] = {
+    other match {
+      case other: CMSZero[_] => this
+      case other: CMSItem[K] => this + other.item
+      case other: SparseCMS[K] =>
+        val newTable = exactCountTable ++ other.exactCountTable
+        if (newTable.size < width * depth) {
+          // still sparse
+          SparseCMS(newTable, totalCount + other.totalCount, params)
+        } else {
+          // Create new CMSInstace
+          var cms = CMSInstance[K](params)
```

Comment:
let's remove this duplication with a `toDense: CMS[K]` method which you can call here and in `+`.